### ISSUE TITLE
Failed jobs are not reactivate after backoff

### ISSUE THREAD

**Issue by deepthidevaki** [camunda/zeebe#14329]
**Describe the bug**

[E2E tests](https://camunda.slack.com/archives/C013MEVQ4M9/p1694702959002779) failed because the send-ping job is not completed. All the instances that did not complete are stuck in `send-message-ping` or `send-message-pong` tasks.

![image](https://github.com/camunda/zeebe/assets/1997478/2b210594-5e5e-47cf-8bae-15a5ba3c0006)

Looking at the logs, publishing message failed due to resource exhausted (there were broker restarts and leader change around this time). So the job was failed with a retry backoff of 30 seconds. However, the worker logs and the exported records show that these jobs were not re-activated. 

Here is screenshot showing exported records in elastic search for some of these jobs.

![tmp-78371-iIZGT5mhN4lx-image](https://github.com/camunda/zeebe/assets/1997478/4dbb026d-31e2-42f2-8d0b-a98e872cc481)

![tmp-78371-p0xwE2yFVwKD-image](https://github.com/camunda/zeebe/assets/1997478/5a38c06e-ff69-4830-a4fd-51cb7d7f5621)

![tmp-78371-23tXS7l4N8jV-image](https://github.com/camunda/zeebe/assets/1997478/512e83ae-291b-45ef-8d78-29e1c1bc9887)

For a job that retried after backoff, you can see those records in ES. But we don't see them for the above jobs. Here is a an example of such a job were retrying after backoff worked successfully (from the same cluster and for same job type). 
![tmp-78371-cK9yR3YYFP5d-image](https://github.com/camunda/zeebe/assets/1997478/dec2c231-cb4a-4f03-ac87-7f94c54fb7b4)

Also ran `zbctl activate jobs "send-message-ping"` locally against this cluster. But it returned  no jobs, even though operate shows these jobs as not completed.

**To Reproduce**

Not sure. But I think restarts and leader change may have something to do with this.

**Expected behavior**

Failed jobs should be re-activatable after backoff timeout.

**Log/Stacktrace**

There were no relevant error in Zeebe. But here are the logs around 14.09.2023 16:33 CEST. All uncompleted tasks were created around 2023-09-14 16:32 - 16:33
[downloaded-logs-20230915-105930.zip](https://github.com/camunda/zeebe/files/12618554/downloaded-logs-20230915-105930.zip)

Here is the snapshot of partition 3 on which these jobs exist. 
[snapshot-partition-3.zip](https://github.com/camunda/zeebe/files/12617504/snapshot-partition-3.zip)

**Environment:**
- Zeebe Version: 8.1.16-SNAPSHOT (8.1.16-SNAPSHOT-stable-8.1-8522cb63)

**Comment by korthout** [comment]
ZPA triage:
- when this happens it's quite bad, those jobs will not picked up again
- we have tests for this, so it must have been a special situation
- may be hard to reproduce
- marking this as `later` unless it's reported again

**Comment by abbasadel** [comment]
ZPA triage:
- We see this issue happening more often now (today's morning and lastweek)
- We don't think this is related to the latest changes since it also happens on 8.0 and 8.3
- We think we should start investigating this and give it a high priority as 'upcoming'

**Comment by deepthidevaki** [comment]
All E2E tests are failing every week due to this bug. It would be good to fix this asap. If the fix is not planned, could you please disable this `ping-pong` process in e2e tests? It doesn't make sense to run the tests if it keeps failing.

**Comment by megglos** [comment]
reached out on slack on this, see https://camunda.slack.com/archives/C037W9NMATG/p1700578334906099

**Comment by nicpuppa** [comment]
From the snapshot provided me and @tmetzke saw that the `JOB_BACKOFF` column family is empty. Possible scenarios:

- Job Failed event is not correctly replayed, after the restart new leader doesn't know about the failed job that needs to be retried
- JOB_BACKOFF column family might be [cleaned up ](https://github.com/camunda/zeebe/blob/1c6c3cc2ec75bdb298c1e644131a586edbf0f501/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java#L241)

**Comment by deepthidevaki** [comment]
> It transitioned to follower in term two, then it should transition to follower on term 3 but it says later it does on term 2

`Transition to FOLLOWER on term 2 completed`
It transitioned to Follower on term 2. 

```
Transition to FOLLOWER on term 3 requested.
13:36:22.451 [Broker-1] [ZeebePartition-1] [zb-actors-1] DEBUG io.camunda.zeebe.broker.system - Partition role transitioning from FOLLOWER to CANDIDATE in term 3
13:36:22.451 [Broker-1] [ZeebePartition-1] [zb-actors-1] INFO  io.camunda.zeebe.broker.system - Prepare transition from FOLLOWER on term 2 to FOLLOWER
```
Then it is "Preparing" transition from Follower on term 2. 

```
Transition to FOLLOWER on term 3 starting
```
Then it is transitioning to term 3.

So everything looks good here.

**Comment by oleschoenburg** [comment]
Hm, I guess this was my bad then :disappointed: https://github.com/camunda/zeebe/commit/1d82e6efc279043b9ff582cf5f66c04afbf53146

I guess the cause is that the record backoff is not supposed to be the point in time where the backoff "expires", i.e. what's in the backoff CF.

**Comment by Zelldon** [comment]
On fail we use the `getRecurringTime` to store the job in the backoff column
```
  @Override
  public void fail(final long key, final JobRecord updatedValue) {
    if (updatedValue.getRetries() > 0) {
      if (updatedValue.getRetryBackoff() > 0) {
        addJobBackoff(key, updatedValue.getRecurringTime());
        updateJob(key, updatedValue, State.FAILED);
      } else {
        updateJob(key, updatedValue, State.ACTIVATABLE);
      }
    } else {
      updateJob(key, updatedValue, State.FAILED);
      makeJobNotActivatable(updatedValue.getTypeBuffer(), updatedValue.getTenantId());
    }
  }
```

Later we check against the backoff of the record. I guess this is the issue?

https://github.com/camunda/zeebe/blob/1c6c3cc2ec75bdb298c1e644131a586edbf0f501/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java#L246

**Comment by Zelldon** [comment]
Thanks @deepthidevaki I'm always (still) confused by these logs.

I guess what I miss is the to term which it is transitioning to like instead of

```
Prepare transition from FOLLOWER on term 2 to FOLLOWER
```

```
Prepare transition from FOLLOWER on term 2 to FOLLOWER on term 3
```

**Comment by Zelldon** [comment]
PR is coming

**Comment by Zelldon** [comment]
I created https://github.com/camunda/zeebe/pull/15419 to fix the general issue.

@camunda/zeebe-process-automation please take a look regarding how to handle already stuck jobs, I would like to hand this over to you from here on.

**Comment by abbasadel** [comment]
Thanks @koevskinikola for mentioning this, I removed the label "version". it seems they were put mistakely by the change-log command.

**Comment by abbasadel** [comment]
As discussed with @nicpuppa , we also need to backport this fix to 8.1, 8.2 plus 8.3
